### PR TITLE
Refactor genomic range sharing into utils-java.

### PR DESCRIPTION
**Pull request by deflaux**
Logic to shard genomic ranges should be all in one place to be reused by both dataflow and spark.  Its also important to ensure that the full collection of shards is shuffled so that parallel requests will be distributed across the dataset.

This pull request is part 1 of many parts to refactor some logic currently in dataflow-java down into utils-java so that it can be more widely re-used.  Since this code may change yet again, sending these pull requests to a branch for now instead of master.

### CODE REVIEW

File: src/main/java/com/google/cloud/genomics/utils/Contig.java [hunk 1]
```diff
@@ -75,23 +85,6 @@ public String toString() {
...
-    }
-    Collections.shuffle(shards); // Shuffle shards for better backend performance
-    return shards;
-  }
-
-  public List<Contig> getShards() {
-    return getShards(DEFAULT_NUMBER_OF_BASES_PER_SHARD);
-  }
-
   public SearchVariantsRequest getVariantsRequest(String variantSetId) {
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: In my opinion this method belongs in some variant-related class, not here. It would be something like an interface that's implemented by a VCF file reader and a Genomics API client.

The benefit is that (i) others can add other implementations if they want (reading from GCS, Hadoop, whatever comes next), and (ii) the contig class changes less (this is good because lots of code will come to depend on contig).

File: src/main/java/com/google/cloud/genomics/utils/Contig.java [hunk 2]
```diff
@@ -100,6 +93,15 @@ public SearchVariantsRequest getVariantsRequest(String variantSetId) {
         .setEnd(end);
   }
 
+  public StreamVariantsRequest getStreamVariantsRequest(String variantSetId) {
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: So instead of adding a new method here, I'd move the original method and place this one along with it.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java [hunk 3]
```diff
@@ -108,7 +110,41 @@ public SearchReadsRequest getReadsRequest(String readGroupSetId) {
         .setEnd(end);
   }
 
-  public static Iterable<Contig> parseContigsFromCommandLine(String contigsArgument) {
+  public StreamReadsRequest getStreamReadsRequest(String readGroupSetId) {
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: Same thing for those, though I understand if you want to do that in a separate PR.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java [hunk 4]
```diff
@@ -108,7 +110,41 @@ public SearchReadsRequest getReadsRequest(String readGroupSetId) {
...
+    return StreamReadsRequest.newBuilder()
+        .setReadGroupSetId(readGroupSetId)
+        .setReferenceName(referenceName)
+        .setStart(start)
+        .setEnd(end)
+        .build();
+  }
+
+  private List<Contig> getShards() {
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: As a matter of taste, I much prefer when all the public methods are first in the file, then all the private ones at the bottom. The reason is that most people are just users of the class, so they can stop reading after the last public method.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java [hunk 5]
```diff
@@ -148,5 +184,78 @@ public Contig apply(String contigString) {
 
     return contigs;
   }
+  
+  /**
+   * Get a list of Contigs representing sharded windows of all genomic regions within the variant set.
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: We're using 4 different words to mean the same thing: contig, shard, region, and interval. Let's try to focus on just one. Unless we want to rename this class, I suggest that our best candidate is "contig". We'll use that to mean "a contiguous region of the genome". 

So this function can be described as (i) loading the specified variants via the API, (ii) casting those variants to contigs - i.e. only keeping the information about where the variant is.

I'm a bit confused about why this should be here. Loading variants via the API has nothing to do with contigs so it doesn't feel like it belongs here. The second part sounds like it should be a method on the variant object: variant.getContig(). Surely that either already exists, or we should add it because knowing where a variant is seems fundamental.

Given this, I don't see why either step belongs here. Dataflow is good at combining DoFns, so it's already easy enough to write variantIDs.apply(new LoadVariantFn()).apply(new CastToContigFn()). No?

File: src/main/java/com/google/cloud/genomics/utils/Contig.java [hunk 6]
```diff
@@ -148,5 +184,78 @@ public Contig apply(String contigString) {
...
+   * 
+   * These are useful for pipelines that operate in parallel against data within a specific variant set.
+   * 
+   * @param genomics - the genomics client
+   * @param variantSetId - the id of the variant set for which to create shards
+   * @param sexChromosomeFilter - whether or not to include the sex chromosomes in the list of shards
+   * @return a shuffled list of shards
+   * @throws IOException
+   */
+  public static List<Contig> getAllShardsInVariantSet(final Genomics genomics, final String variantSetId,
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: I would not include this form of the function. It's not needed in a minimal API, and adds work for us. Same for all other variants that pick DEFAULT_NUMBER_OF_BASES_PER_SHARD instead of getting it via argument.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java [hunk 7]
```diff
@@ -148,5 +184,78 @@ public Contig apply(String contigString) {
...
+      final SexChromosomeFilter sexChromosomeFilter) throws IOException {
+    return getAllShardsInVariantSet(genomics, variantSetId, sexChromosomeFilter, DEFAULT_NUMBER_OF_BASES_PER_SHARD); 
+  }
+  
+  /**
+   * Get a list of Contigs representing sharded windows of all genomic regions within the variant set.
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: This doesn't tell the whole story - unless I'm mistaken it looks like those contigs start at 0, instead of starting at the variant's start position (line 182). So this isn't actually a simple typecast from Variant to bounding interval.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java [hunk 8]
```diff
@@ -148,5 +184,78 @@ public Contig apply(String contigString) {
...
+  /**
+   * Get a list of Contigs representing sharded windows of the specified genomic regions expressed in the command line argument.
+   * 
+   * These are useful for pipelines that operate in parallel against data within a specific variant set or read group set.
+   * 
+   * @param contigsArgument - the command line parameter expressing the specified genomic regions
+   *                            format is chromosome:start:end[,chromosome:start:end]
+   * @return a shuffled list of shards
+   */
+  public static List<Contig> getSpecifiedShards(final String contigsArgument) {
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: Let's agree on a single word to mean a genomic interval. I think that word is contig. So this function might be renamed e.g. to "parse", as in "Contig.parse("chr1:1:100")". Or "of", a name Dataflow often uses for static constructor-like methods.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java [hunk 9]
```diff
@@ -148,5 +184,78 @@ public Contig apply(String contigString) {
...
+   * Get a list of Contigs representing sharded windows of the specified genomic regions expressed in the command line argument.
+   * 
+   * These are useful for pipelines that operate in parallel against data within a specific variant set or read group set.
+   * 
+   * @param contigsArgument - the command line parameter expressing the specified genomic regions
+   *                            format is chromosome:start:end[,chromosome:start:end]
+   * @param numberOfBasesPerShard - the maximum size of the shard window in terms of number of bases
+   * @return a shuffled list of shards
+   */
+  public static List<Contig> getSpecifiedShards(final String contigsArgument, final long numberOfBasesPerShard) {
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: Let's agree on a single word to mean a genomic interval. I think that word is contig. So this function might be renamed e.g. to "parse", as in "Contig.parse("chr1:1:100")". Or "of", a name Dataflow often uses for static constructor-like methods.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java (hunk 5 above)
Comment (deflaux) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: This doesn't load the variants.

It just determines what contigs a variant set has and then also shards them.  It comes into play when you use command line argument `--allReferences` via spark or dataflow.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java (hunk 1 above)
Comment (deflaux) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: Agreed.  This will happen in a separate PR.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java (hunk 4 above)
Comment (deflaux) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: Will do!

File: src/main/java/com/google/cloud/genomics/utils/Contig.java (hunk 6 above)
Comment (deflaux) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: Will remove.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java (hunk 8 above)
Comment (deflaux) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: Will remove.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java (hunk 7 above)
Comment (deflaux) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: variantsets.get for v1beta2 only returns the upper bound per reference, hence the hard-coded lower bound

File: src/main/java/com/google/cloud/genomics/utils/Contig.java (hunk 5 above)
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: You're right, I was confused. This also explains why we spit out contigs that start at 0 - that's because it's what the API gives us. I still would like to see all those Genomic API-dependent features out of the contig class and into some helper class, but I'll grant you this is a matter of taste.

File: src/test/java/com/google/cloud/genomics/utils/ContigTest.java [hunk 10]
```diff
@@ -15,43 +15,76 @@
...
-import static org.junit.Assert.assertEquals;
+import com.google.api.services.genomics.model.SearchReadsRequest;
+import com.google.api.services.genomics.model.SearchVariantsRequest;
+import com.google.common.base.Function;
+import com.google.common.base.Joiner;
+import com.google.common.collect.Ordering;
 
 @RunWith(JUnit4.class)
 public class ContigTest {
+
```
Comment (jean-philippe-martin) [inline src/test/java/com/google/cloud/genomics/utils/ContigTest.java]: I see you're not testing all of the new functions. It would be nice if we did.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java (hunk 9 above)
Comment (deflaux) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: Agreed on using fewer synonyms.  There are times though that we have both contigs and shards.  Take a look at the example below and let me know what you think?

* https://github.com/deflaux/dataflow-java/blob/refactor-part1/src/main/java/com/google/cloud/genomics/dataflow/pipelines/CountReads.java#L182
* https://github.com/deflaux/dataflow-java/blob/refactor-part1/src/main/java/com/google/cloud/genomics/dataflow/pipelines/CountReads.java#L195 the sharding happens in a different class

File: src/test/java/com/google/cloud/genomics/utils/ContigTest.java (hunk 10 above)
Comment (deflaux) [inline src/test/java/com/google/cloud/genomics/utils/ContigTest.java]: The new methods that copy Contig fields into request fields aren't that interesting to test.

The test for the new method `getSpecifiedShards` exercises it and the private methods upon which it depends.

There is no test for the new method `getAllShardsInVariantSet` because it just makes an api call and then exercises the same logic as is used by `getSpecifiedShards`.  I could mock it out or add an integration test.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java (hunk 7 above)
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: Indeed, so the comment should reflect that reality.

**Comment by deflaux** [comment]
@jean-philippe-martin PTAL

* rearranged code per feedback and added integration tests
* also refactored read sharding from dataflow-java to utils-java
* please be sure to view https://github.com/googlegenomics/dataflow-java/pull/116 so that you can see where much of this code came from

Happy to rename anything to clean up/consolidate terminology.

File: src/main/java/com/google/cloud/genomics/utils/Contig.java (hunk 9 above)
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Contig.java]: Agreed, that code uses both terms. I'm not asking you to go change it, but having seen those terms a few times I wanted to start the conversation - if we reach an agreement then the big rename can be a separate CL.

File: src/main/java/com/google/cloud/genomics/utils/GenomicsUtils.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+import com.google.common.collect.Lists;
+
+/**
+ * Convenience routines for fetching ids in the hierarchy of data within the Genomics API and other data lookups.
+ */
+public class GenomicsUtils {
+  
+  /**
+   * Gets ReadGroupSetIds from a given datasetId using the Genomics API.
+   */
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/GenomicsUtils.java]: perhaps add: "Throws an exception if there is none", since this behavior may be surprising to some (reading from files or console doesn't throw if the input is empty).

File: src/main/java/com/google/cloud/genomics/utils/GenomicsUtils.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+      output.add(r.getId());
+    }
+    if (output.isEmpty()) {
+      throw new IOException("Dataset " + datasetId + " does not contain any ReadGroupSets");
+    }
+    return output;
+  }
+  
+  /**
+   * Gets VariantSetIds from a given datasetId using the Genomics API.
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/GenomicsUtils.java]: mention the exception here, and in the next method too.

File: src/main/java/com/google/cloud/genomics/utils/Paginator.java [hunk 13]
```diff
@@ -94,9 +94,9 @@
  *String datasetId = ...;
  *Paginator.Readsets searchReadsets = Paginator.Readsets.create(stub);
  *for (Readset readset =
- *    searchReadsets.search(new SearchReadsetsRequest().setDatasetId(datasetId))) {
+ *    searchReadsets.search(new SearchReadsetsRequest().setDatasetId(datasetId))) &#123;
  *  // do something with readset
- *}
+ *&#125;
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Paginator.java]: did we mean to do this? Is it necessary for something? It certainly makes the comment harder to read in the IDE.

File: src/main/java/com/google/cloud/genomics/utils/ShardUtils.java [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+ *      be shuffled for the next step could be quite large, if a combiner is not used.
+ *  - Shard strategies can also differ based on paginated vs. streaming APIs.
+ *  
+ * This is clearly a work in progress, many of the issues described above are not addressed
+ * in the code below.  But let's consolidate the tribal knowledge here for best practices
+ * in manual sharding for reuse by Spark or any other systems processing data in parallel.
+ * 
+ * For Dataflow, custom sources should be preferred over this manual sharding approach.
+ *
+ */
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/ShardUtils.java]: excellent comment!

File: src/test/java/com/google/cloud/genomics/utils/IntegrationTestHelper.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    new ReferenceBound().setReferenceName("chrM").setUpperBound(1000001L),
+    new ReferenceBound().setReferenceName("chrX").setUpperBound(156231278L),
+    new ReferenceBound().setReferenceName("chrY").setUpperBound(60032946L)
+  };
+
+  // Test configuration constants
+  public final String API_KEY = System.getenv("GOOGLE_API_KEY");
+  public final OfflineAuth auth;
+  
+  public IntegrationTestHelper() throws IOException, GeneralSecurityException {
```
Comment (jean-philippe-martin) [inline src/test/java/com/google/cloud/genomics/utils/IntegrationTestHelper.java]: I like the idea of a test helper. Here there's a risk that users may forget to call the constructor - all methods are static after all, and it's all too easy to write IntegreationTestHelper.API_KEY.

What I'd like to suggest instead is to have either
(a) API_KEY and auth are protected, only accessible via non-static getters. Thus it's impossible to forget to call the constructor (if we need it)
(b) (I prefer this one) API_KEY and auth are static protected, only accessible via static getters. The getters will grab the information on the first call, stash it for later calls. The upside is that if we don't need auth but need something else from the helper, we can still use it without getting an error. The downside is this isn't thread-safe so we either live with that or spend a bit more time making it so.

File: src/test/java/com/google/cloud/genomics/utils/ShardUtilsITCase.java [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+        new Contig("chr8", 0L, 147299246L)
+        .getStreamVariantsRequest(helper.PLATINUM_GENOMES_VARIANTSET),
+        new Contig("chr9", 0L, 142027288L)
+        .getStreamVariantsRequest(helper.PLATINUM_GENOMES_VARIANTSET),
+        new Contig("chrM", 0L, 1000001L)
+        .getStreamVariantsRequest(helper.PLATINUM_GENOMES_VARIANTSET)
+    };
+    
+    // These shards are "too big" to use in practice but for this test it keeps the
+    // expected result from getting crazy long.
```
Comment (jean-philippe-martin) [inline src/test/java/com/google/cloud/genomics/utils/ShardUtilsITCase.java]: Good call :)

**Comment by jean-philippe-martin** [comment]
Thank you! LGTM, aside from minor comments. Feel free to merge once ready.

File: src/main/java/com/google/cloud/genomics/utils/Paginator.java (hunk 13 above)
Comment (deflaux) [inline src/main/java/com/google/cloud/genomics/utils/Paginator.java]: Yes, it was to clear an eclipse error on javadoc formatting.

Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/utils/Paginator.java]: I checked out [stackoverflow](http://stackoverflow.com/questions/647195/how-do-you-escape-curly-braces-in-javadoc-inline-tags-such-as-the-code-tag) for a bit of background information, and man is this messed up.
 * `<code>` requires you to escape angle brackets
 * `{@code ... }` requires you to escape braces

**Comment by deflaux** [comment]
Javadocs added.  Will update `IntegrationTestHelper.API_KEY` in both utils-java and dataflow-java in the next set of pull requests.
